packages/cosmic-proto/src/address-hooks.js

```diff
@@ -0,0 +1,231 @@
+  hookBuf.set(bytes, magicLength);
+  hookBuf.set(hookData, magicLength + b);
+
+  // Append the big-endian address length.
+  let len = b;
+  for (let i = 0; i < BASE_ADDRESS_LENGTH_BYTES; i += 1) {
+    hookBuf[hookBuf.length - 1 - i] = len & 0xff;
+    len >>>= 8;
+  }
+
+  return encodeBech32(prefix, hookBuf, charLimit);
+};
+
+/**
+ * @param {string} baseAddress
+ * @param {HookQuery} query
+ * @param {number} [charLimit]
+ */
+export const encodeAddressHook = (baseAddress, query, charLimit) => {
+  const queryStr = queryString.stringify(query);
+
+  const te = new TextEncoder();
+  const hookData = te.encode(`?${queryStr}`);
+  return joinHookedAddress(baseAddress, hookData, charLimit);
+};
+
+/**
+ * @param {string} addressHook
+ * @param {number} [charLimit]
+ * @returns {{ baseAddress: string; query: HookQuery }}
+ */
+export const decodeAddressHook = (addressHook, charLimit) => {
+  const { baseAddress, hookData } = splitHookedAddress(addressHook, charLimit);
+  const hookStr = new TextDecoder().decode(hookData);
+  if (hookStr && !hookStr.startsWith('?')) {
+    throw Error(`Hook data does not start with '?': ${hookStr}`);
+  }
+
+  /** @type {HookQuery} */
+  const query = queryString.parse(hookStr);
+  return { baseAddress, query };
+};
+
+/**
+ * @param {string} specimen
+ * @param {number} [charLimit]
+ * @returns {string | { baseAddress: string; hookData: Uint8Array }}
+ */
+export const splitHookedAddressUnsafe = (
+  specimen,
+  charLimit = DEFAULT_HOOKED_ADDRESS_CHAR_LIMIT,
+) => {
+  const { prefix, bytes } = decodeBech32(specimen, charLimit);
+
+  const magicLength = ADDRESS_HOOK_MAGIC.length;
+  for (let i = 0; i < magicLength; i += 1) {
+    if (bytes[i] !== ADDRESS_HOOK_MAGIC[i]) {
+      return { baseAddress: specimen, hookData: new Uint8Array() };
+    }
+  }
+
+  let len = 0;
+  for (let i = BASE_ADDRESS_LENGTH_BYTES - 1; i >= 0; i -= 1) {
+    const byte = bytes.at(-i - 1);
+    if (byte === undefined) {
+      return `Cannot get base address length from byte ${-i - 1} of ${bytes.length}`;
+    }
+    len <<= 8;
+    len |= byte;
+  }
+
+  const b = len;
+  if (b > bytes.length - BASE_ADDRESS_LENGTH_BYTES - magicLength) {
+    return `Base address length 0x${b.toString(16)} is longer than specimen length ${bytes.length - BASE_ADDRESS_LENGTH_BYTES - magicLength}`;
+  }
+
+  const baseAddressBuf = bytes.subarray(magicLength, magicLength + b);
+  const baseAddress = encodeBech32(prefix, baseAddressBuf, charLimit);
+
+  const hookData = bytes.subarray(magicLength + b, -BASE_ADDRESS_LENGTH_BYTES);
+
+  return { baseAddress, hookData };
+};
+
+/**
+ * @param {string} specimen
+ * @param {number} [charLimit]
+ * @returns {{
+ *   baseAddress: string;
+ *   hookData: Uint8Array;
+ * }}
+ */
+export const splitHookedAddress = (specimen, charLimit) => {
+  const result = splitHookedAddressUnsafe(specimen, charLimit);
+  if (typeof result === 'object') {
+    return result;
+  }
+  throw Error(result);
```

Comment:
kind of an obscure diagnostic

Reply:
ah... I misread. I was thinking that the `splitHookedAddressUnsafe` returned the specimen.